<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.10.0" />
<title>mgcplotter.genbank API documentation</title>
<meta name="description" content="" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>mgcplotter.genbank</code></h1>
</header>
<section id="section-intro">
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">from functools import cached_property
from pathlib import Path
from typing import Any, List, Optional, Union

from Bio import SeqIO, SeqUtils
from Bio.SeqFeature import FeatureLocation, Seq, SeqFeature
from Bio.SeqRecord import SeqRecord


class Genbank:
    &#34;&#34;&#34;Genbank Class&#34;&#34;&#34;

    def __init__(
        self,
        gbk_file: Union[str, Path],
        name: str = &#34;&#34;,
    ):
        &#34;&#34;&#34;Constructor

        Args:
            gbk_file (Union[str, StringIO, Path]): Genbank file
            name (str, optional): Name
        &#34;&#34;&#34;
        self.gbk_file: Path = Path(gbk_file)
        self.name: str = name if name != &#34;&#34; else self.gbk_file.with_suffix(&#34;&#34;).name
        self._records: List[SeqRecord] = list(SeqIO.parse(gbk_file, &#34;genbank&#34;))

    @property
    def genome_length(self) -&gt; int:
        &#34;&#34;&#34;Genome sequence length&#34;&#34;&#34;
        return len(self.genome_seq)

    @property
    def genome_seq(self) -&gt; str:
        &#34;&#34;&#34;Genome sequence (join all contig sequences)&#34;&#34;&#34;
        return &#34;&#34;.join(self.contig_seqs)

    @property
    def contig_seqs(self) -&gt; List[str]:
        &#34;&#34;&#34;Contig sequences&#34;&#34;&#34;
        return [str(r.seq) for r in self._records]

    @cached_property
    def average_gc(self) -&gt; float:
        &#34;&#34;&#34;Average GC content&#34;&#34;&#34;
        return SeqUtils.GC(self.genome_seq)

    def gc_skew(self, window_size: int = 5000, step_size: int = 2000) -&gt; List[float]:
        &#34;&#34;&#34;Calculate GC skew in sliding window

        Args:
            window_size (int, optional): Window size
            step_size (int, optional): Step size

        Returns:
            List[float]: GC skew values in sliding window
        &#34;&#34;&#34;
        gc_skew_values = []
        seq = self.genome_seq
        for i in range(0, len(seq), step_size):
            start_pos = i - int(window_size / 2)
            start_pos = 0 if start_pos &lt; 0 else start_pos
            end_pos = i + int(window_size / 2)
            end_pos = len(seq) if end_pos &gt; len(seq) else end_pos

            subseq = seq[start_pos:end_pos]
            g = subseq.count(&#34;G&#34;) + subseq.count(&#34;g&#34;)
            c = subseq.count(&#34;C&#34;) + subseq.count(&#34;c&#34;)
            try:
                skew = (g - c) / float(g + c)
            except ZeroDivisionError:
                skew = 0.0
            gc_skew_values.append(skew)
        return gc_skew_values

    def gc_content(self, window_size: int = 5000, step_size: int = 2000) -&gt; List[float]:
        &#34;&#34;&#34;Calculate GC content in sliding window

        Args:
            window_size (int, optional): Window size
            step_size (int, optional): Step size

        Returns:
            List[float]: GC content values in sliding window
        &#34;&#34;&#34;
        gc_content_values = []
        seq = self.genome_seq
        for i in range(0, len(seq), step_size):
            start_pos = i - int(window_size / 2)
            start_pos = 0 if start_pos &lt; 0 else start_pos
            end_pos = i + int(window_size / 2)
            end_pos = len(seq) if end_pos &gt; len(seq) else end_pos

            subseq = seq[start_pos:end_pos]
            gc_content_values.append(SeqUtils.GC(subseq))
        return gc_content_values

    def extract_all_features(
        self,
        feature_type: str = &#34;CDS&#34;,
        target_strand: Optional[int] = None,
    ) -&gt; List[SeqFeature]:
        &#34;&#34;&#34;Extract all features

        Args:
            feature_type (str): Feature type to extract
            target_strand (Optional[int]): Target starnd to extract
        Returns:
            List[SeqFeature]: All features
        &#34;&#34;&#34;
        extract_features = []
        base_len = 0
        for record in self._records:
            features = [f for f in record.features if f.type == feature_type]
            for f in features:
                if feature_type == &#34;CDS&#34;:
                    # Exclude pseudogene (no translated gene)
                    translation = f.qualifiers.get(&#34;translation&#34;, [None])[0]
                    if translation is None:
                        continue
                start = self._to_int(f.location.parts[0].start) + base_len
                end = self._to_int(f.location.parts[-1].end) + base_len
                # Exclude feature that straddle start position
                if start &gt; end:
                    continue
                # Extract only target strand feature
                if target_strand is not None and f.strand != target_strand:
                    continue

                extract_features.append(
                    SeqFeature(
                        location=FeatureLocation(start, end, f.strand),
                        type=f.type,
                        qualifiers=f.qualifiers,
                    ),
                )
            base_len += len(record.seq)

        return extract_features

    def write_cds_fasta(
        self,
        fasta_outfile: Union[str, Path],
    ):
        &#34;&#34;&#34;Write CDS protein features fasta file

        Args:
            fasta_outfile (Union[str, Path]): _description_
        &#34;&#34;&#34;
        features = self.extract_all_features(&#34;CDS&#34;, None)
        cds_seq_records: List[SeqRecord] = []
        for idx, feature in enumerate(features, 1):
            qualifiers = feature.qualifiers
            protein_id = qualifiers.get(&#34;protein_id&#34;, [None])[0]
            product = qualifiers.get(&#34;product&#34;, [&#34;&#34;])[0]
            translation = qualifiers.get(&#34;translation&#34;, [None])[0]

            start = self._to_int(feature.location.start)
            end = self._to_int(feature.location.end)
            strand = &#34;+&#34; if feature.strand == 1 else &#34;-&#34;

            location_id = f&#34;|{start}_{end}_{strand}|&#34;
            if protein_id is None:
                seq_id = f&#34;GENE{idx:06d}{location_id}&#34;
            else:
                seq_id = f&#34;GENE{idx:06d}_{protein_id}{location_id}&#34;

            cds_seq_record = SeqRecord(
                seq=Seq(translation), id=seq_id, description=product
            )
            cds_seq_records.append(cds_seq_record)

        SeqIO.write(cds_seq_records, fasta_outfile, &#34;fasta-2line&#34;)

    def write_genome_fasta(
        self,
        outfile: Union[str, Path],
    ) -&gt; None:
        &#34;&#34;&#34;Write genome fasta file

        Args:
            outfile (Union[str, Path]): Output genome fasta file
        &#34;&#34;&#34;
        write_seq = self.genome_seq
        with open(outfile, &#34;w&#34;) as f:
            f.write(f&#34;&gt;{self.name}\n{write_seq}\n&#34;)

    def _to_int(self, value: Any) -&gt; int:
        &#34;&#34;&#34;Convert to int (Required for AbstractPostion|ExactPostion)&#34;&#34;&#34;
        return int(str(value).replace(&#34;&lt;&#34;, &#34;&#34;).replace(&#34;&gt;&#34;, &#34;&#34;))</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-classes">Classes</h2>
<dl>
<dt id="mgcplotter.genbank.Genbank"><code class="flex name class">
<span>class <span class="ident">Genbank</span></span>
<span>(</span><span>gbk_file: Union[str, pathlib.Path], name: str = '')</span>
</code></dt>
<dd>
<div class="desc"><p>Genbank Class</p>
<p>Constructor</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>gbk_file</code></strong> :&ensp;<code>Union[str, StringIO, Path]</code></dt>
<dd>Genbank file</dd>
<dt><strong><code>name</code></strong> :&ensp;<code>str</code>, optional</dt>
<dd>Name</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class Genbank:
    &#34;&#34;&#34;Genbank Class&#34;&#34;&#34;

    def __init__(
        self,
        gbk_file: Union[str, Path],
        name: str = &#34;&#34;,
    ):
        &#34;&#34;&#34;Constructor

        Args:
            gbk_file (Union[str, StringIO, Path]): Genbank file
            name (str, optional): Name
        &#34;&#34;&#34;
        self.gbk_file: Path = Path(gbk_file)
        self.name: str = name if name != &#34;&#34; else self.gbk_file.with_suffix(&#34;&#34;).name
        self._records: List[SeqRecord] = list(SeqIO.parse(gbk_file, &#34;genbank&#34;))

    @property
    def genome_length(self) -&gt; int:
        &#34;&#34;&#34;Genome sequence length&#34;&#34;&#34;
        return len(self.genome_seq)

    @property
    def genome_seq(self) -&gt; str:
        &#34;&#34;&#34;Genome sequence (join all contig sequences)&#34;&#34;&#34;
        return &#34;&#34;.join(self.contig_seqs)

    @property
    def contig_seqs(self) -&gt; List[str]:
        &#34;&#34;&#34;Contig sequences&#34;&#34;&#34;
        return [str(r.seq) for r in self._records]

    @cached_property
    def average_gc(self) -&gt; float:
        &#34;&#34;&#34;Average GC content&#34;&#34;&#34;
        return SeqUtils.GC(self.genome_seq)

    def gc_skew(self, window_size: int = 5000, step_size: int = 2000) -&gt; List[float]:
        &#34;&#34;&#34;Calculate GC skew in sliding window

        Args:
            window_size (int, optional): Window size
            step_size (int, optional): Step size

        Returns:
            List[float]: GC skew values in sliding window
        &#34;&#34;&#34;
        gc_skew_values = []
        seq = self.genome_seq
        for i in range(0, len(seq), step_size):
            start_pos = i - int(window_size / 2)
            start_pos = 0 if start_pos &lt; 0 else start_pos
            end_pos = i + int(window_size / 2)
            end_pos = len(seq) if end_pos &gt; len(seq) else end_pos

            subseq = seq[start_pos:end_pos]
            g = subseq.count(&#34;G&#34;) + subseq.count(&#34;g&#34;)
            c = subseq.count(&#34;C&#34;) + subseq.count(&#34;c&#34;)
            try:
                skew = (g - c) / float(g + c)
            except ZeroDivisionError:
                skew = 0.0
            gc_skew_values.append(skew)
        return gc_skew_values

    def gc_content(self, window_size: int = 5000, step_size: int = 2000) -&gt; List[float]:
        &#34;&#34;&#34;Calculate GC content in sliding window

        Args:
            window_size (int, optional): Window size
            step_size (int, optional): Step size

        Returns:
            List[float]: GC content values in sliding window
        &#34;&#34;&#34;
        gc_content_values = []
        seq = self.genome_seq
        for i in range(0, len(seq), step_size):
            start_pos = i - int(window_size / 2)
            start_pos = 0 if start_pos &lt; 0 else start_pos
            end_pos = i + int(window_size / 2)
            end_pos = len(seq) if end_pos &gt; len(seq) else end_pos

            subseq = seq[start_pos:end_pos]
            gc_content_values.append(SeqUtils.GC(subseq))
        return gc_content_values

    def extract_all_features(
        self,
        feature_type: str = &#34;CDS&#34;,
        target_strand: Optional[int] = None,
    ) -&gt; List[SeqFeature]:
        &#34;&#34;&#34;Extract all features

        Args:
            feature_type (str): Feature type to extract
            target_strand (Optional[int]): Target starnd to extract
        Returns:
            List[SeqFeature]: All features
        &#34;&#34;&#34;
        extract_features = []
        base_len = 0
        for record in self._records:
            features = [f for f in record.features if f.type == feature_type]
            for f in features:
                if feature_type == &#34;CDS&#34;:
                    # Exclude pseudogene (no translated gene)
                    translation = f.qualifiers.get(&#34;translation&#34;, [None])[0]
                    if translation is None:
                        continue
                start = self._to_int(f.location.parts[0].start) + base_len
                end = self._to_int(f.location.parts[-1].end) + base_len
                # Exclude feature that straddle start position
                if start &gt; end:
                    continue
                # Extract only target strand feature
                if target_strand is not None and f.strand != target_strand:
                    continue

                extract_features.append(
                    SeqFeature(
                        location=FeatureLocation(start, end, f.strand),
                        type=f.type,
                        qualifiers=f.qualifiers,
                    ),
                )
            base_len += len(record.seq)

        return extract_features

    def write_cds_fasta(
        self,
        fasta_outfile: Union[str, Path],
    ):
        &#34;&#34;&#34;Write CDS protein features fasta file

        Args:
            fasta_outfile (Union[str, Path]): _description_
        &#34;&#34;&#34;
        features = self.extract_all_features(&#34;CDS&#34;, None)
        cds_seq_records: List[SeqRecord] = []
        for idx, feature in enumerate(features, 1):
            qualifiers = feature.qualifiers
            protein_id = qualifiers.get(&#34;protein_id&#34;, [None])[0]
            product = qualifiers.get(&#34;product&#34;, [&#34;&#34;])[0]
            translation = qualifiers.get(&#34;translation&#34;, [None])[0]

            start = self._to_int(feature.location.start)
            end = self._to_int(feature.location.end)
            strand = &#34;+&#34; if feature.strand == 1 else &#34;-&#34;

            location_id = f&#34;|{start}_{end}_{strand}|&#34;
            if protein_id is None:
                seq_id = f&#34;GENE{idx:06d}{location_id}&#34;
            else:
                seq_id = f&#34;GENE{idx:06d}_{protein_id}{location_id}&#34;

            cds_seq_record = SeqRecord(
                seq=Seq(translation), id=seq_id, description=product
            )
            cds_seq_records.append(cds_seq_record)

        SeqIO.write(cds_seq_records, fasta_outfile, &#34;fasta-2line&#34;)

    def write_genome_fasta(
        self,
        outfile: Union[str, Path],
    ) -&gt; None:
        &#34;&#34;&#34;Write genome fasta file

        Args:
            outfile (Union[str, Path]): Output genome fasta file
        &#34;&#34;&#34;
        write_seq = self.genome_seq
        with open(outfile, &#34;w&#34;) as f:
            f.write(f&#34;&gt;{self.name}\n{write_seq}\n&#34;)

    def _to_int(self, value: Any) -&gt; int:
        &#34;&#34;&#34;Convert to int (Required for AbstractPostion|ExactPostion)&#34;&#34;&#34;
        return int(str(value).replace(&#34;&lt;&#34;, &#34;&#34;).replace(&#34;&gt;&#34;, &#34;&#34;))</code></pre>
</details>
<h3>Instance variables</h3>
<dl>
<dt id="mgcplotter.genbank.Genbank.genome_length"><code class="name">var <span class="ident">genome_length</span> : int</code></dt>
<dd>
<div class="desc"><p>Genome sequence length</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">@property
def genome_length(self) -&gt; int:
    &#34;&#34;&#34;Genome sequence length&#34;&#34;&#34;
    return len(self.genome_seq)</code></pre>
</details>
</dd>
<dt id="mgcplotter.genbank.Genbank.genome_seq"><code class="name">var <span class="ident">genome_seq</span> : str</code></dt>
<dd>
<div class="desc"><p>Genome sequence (join all contig sequences)</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">@property
def genome_seq(self) -&gt; str:
    &#34;&#34;&#34;Genome sequence (join all contig sequences)&#34;&#34;&#34;
    return &#34;&#34;.join(self.contig_seqs)</code></pre>
</details>
</dd>
<dt id="mgcplotter.genbank.Genbank.contig_seqs"><code class="name">var <span class="ident">contig_seqs</span> : List[str]</code></dt>
<dd>
<div class="desc"><p>Contig sequences</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">@property
def contig_seqs(self) -&gt; List[str]:
    &#34;&#34;&#34;Contig sequences&#34;&#34;&#34;
    return [str(r.seq) for r in self._records]</code></pre>
</details>
</dd>
<dt id="mgcplotter.genbank.Genbank.average_gc"><code class="name">var <span class="ident">average_gc</span></code></dt>
<dd>
<div class="desc"><p>Average GC content</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def __get__(self, instance, owner=None):
    if instance is None:
        return self
    if self.attrname is None:
        raise TypeError(
            &#34;Cannot use cached_property instance without calling __set_name__ on it.&#34;)
    try:
        cache = instance.__dict__
    except AttributeError:  # not all objects have __dict__ (e.g. class defines slots)
        msg = (
            f&#34;No &#39;__dict__&#39; attribute on {type(instance).__name__!r} &#34;
            f&#34;instance to cache {self.attrname!r} property.&#34;
        )
        raise TypeError(msg) from None
    val = cache.get(self.attrname, _NOT_FOUND)
    if val is _NOT_FOUND:
        with self.lock:
            # check if another thread filled cache while we awaited lock
            val = cache.get(self.attrname, _NOT_FOUND)
            if val is _NOT_FOUND:
                val = self.func(instance)
                try:
                    cache[self.attrname] = val
                except TypeError:
                    msg = (
                        f&#34;The &#39;__dict__&#39; attribute on {type(instance).__name__!r} instance &#34;
                        f&#34;does not support item assignment for caching {self.attrname!r} property.&#34;
                    )
                    raise TypeError(msg) from None
    return val</code></pre>
</details>
</dd>
</dl>
<h3>Methods</h3>
<dl>
<dt id="mgcplotter.genbank.Genbank.gc_skew"><code class="name flex">
<span>def <span class="ident">gc_skew</span></span>(<span>self, window_size: int = 5000, step_size: int = 2000) ‑> List[float]</span>
</code></dt>
<dd>
<div class="desc"><p>Calculate GC skew in sliding window</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>window_size</code></strong> :&ensp;<code>int</code>, optional</dt>
<dd>Window size</dd>
<dt><strong><code>step_size</code></strong> :&ensp;<code>int</code>, optional</dt>
<dd>Step size</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>List[float]</code></dt>
<dd>GC skew values in sliding window</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def gc_skew(self, window_size: int = 5000, step_size: int = 2000) -&gt; List[float]:
    &#34;&#34;&#34;Calculate GC skew in sliding window

    Args:
        window_size (int, optional): Window size
        step_size (int, optional): Step size

    Returns:
        List[float]: GC skew values in sliding window
    &#34;&#34;&#34;
    gc_skew_values = []
    seq = self.genome_seq
    for i in range(0, len(seq), step_size):
        start_pos = i - int(window_size / 2)
        start_pos = 0 if start_pos &lt; 0 else start_pos
        end_pos = i + int(window_size / 2)
        end_pos = len(seq) if end_pos &gt; len(seq) else end_pos

        subseq = seq[start_pos:end_pos]
        g = subseq.count(&#34;G&#34;) + subseq.count(&#34;g&#34;)
        c = subseq.count(&#34;C&#34;) + subseq.count(&#34;c&#34;)
        try:
            skew = (g - c) / float(g + c)
        except ZeroDivisionError:
            skew = 0.0
        gc_skew_values.append(skew)
    return gc_skew_values</code></pre>
</details>
</dd>
<dt id="mgcplotter.genbank.Genbank.gc_content"><code class="name flex">
<span>def <span class="ident">gc_content</span></span>(<span>self, window_size: int = 5000, step_size: int = 2000) ‑> List[float]</span>
</code></dt>
<dd>
<div class="desc"><p>Calculate GC content in sliding window</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>window_size</code></strong> :&ensp;<code>int</code>, optional</dt>
<dd>Window size</dd>
<dt><strong><code>step_size</code></strong> :&ensp;<code>int</code>, optional</dt>
<dd>Step size</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>List[float]</code></dt>
<dd>GC content values in sliding window</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def gc_content(self, window_size: int = 5000, step_size: int = 2000) -&gt; List[float]:
    &#34;&#34;&#34;Calculate GC content in sliding window

    Args:
        window_size (int, optional): Window size
        step_size (int, optional): Step size

    Returns:
        List[float]: GC content values in sliding window
    &#34;&#34;&#34;
    gc_content_values = []
    seq = self.genome_seq
    for i in range(0, len(seq), step_size):
        start_pos = i - int(window_size / 2)
        start_pos = 0 if start_pos &lt; 0 else start_pos
        end_pos = i + int(window_size / 2)
        end_pos = len(seq) if end_pos &gt; len(seq) else end_pos

        subseq = seq[start_pos:end_pos]
        gc_content_values.append(SeqUtils.GC(subseq))
    return gc_content_values</code></pre>
</details>
</dd>
<dt id="mgcplotter.genbank.Genbank.extract_all_features"><code class="name flex">
<span>def <span class="ident">extract_all_features</span></span>(<span>self, feature_type: str = 'CDS', target_strand: Optional[int] = None) ‑> List[Bio.SeqFeature.SeqFeature]</span>
</code></dt>
<dd>
<div class="desc"><p>Extract all features</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>feature_type</code></strong> :&ensp;<code>str</code></dt>
<dd>Feature type to extract</dd>
<dt><strong><code>target_strand</code></strong> :&ensp;<code>Optional[int]</code></dt>
<dd>Target starnd to extract</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>List[SeqFeature]</code></dt>
<dd>All features</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def extract_all_features(
    self,
    feature_type: str = &#34;CDS&#34;,
    target_strand: Optional[int] = None,
) -&gt; List[SeqFeature]:
    &#34;&#34;&#34;Extract all features

    Args:
        feature_type (str): Feature type to extract
        target_strand (Optional[int]): Target starnd to extract
    Returns:
        List[SeqFeature]: All features
    &#34;&#34;&#34;
    extract_features = []
    base_len = 0
    for record in self._records:
        features = [f for f in record.features if f.type == feature_type]
        for f in features:
            if feature_type == &#34;CDS&#34;:
                # Exclude pseudogene (no translated gene)
                translation = f.qualifiers.get(&#34;translation&#34;, [None])[0]
                if translation is None:
                    continue
            start = self._to_int(f.location.parts[0].start) + base_len
            end = self._to_int(f.location.parts[-1].end) + base_len
            # Exclude feature that straddle start position
            if start &gt; end:
                continue
            # Extract only target strand feature
            if target_strand is not None and f.strand != target_strand:
                continue

            extract_features.append(
                SeqFeature(
                    location=FeatureLocation(start, end, f.strand),
                    type=f.type,
                    qualifiers=f.qualifiers,
                ),
            )
        base_len += len(record.seq)

    return extract_features</code></pre>
</details>
</dd>
<dt id="mgcplotter.genbank.Genbank.write_cds_fasta"><code class="name flex">
<span>def <span class="ident">write_cds_fasta</span></span>(<span>self, fasta_outfile: Union[str, pathlib.Path])</span>
</code></dt>
<dd>
<div class="desc"><p>Write CDS protein features fasta file</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>fasta_outfile</code></strong> :&ensp;<code>Union[str, Path]</code></dt>
<dd><em>description</em></dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def write_cds_fasta(
    self,
    fasta_outfile: Union[str, Path],
):
    &#34;&#34;&#34;Write CDS protein features fasta file

    Args:
        fasta_outfile (Union[str, Path]): _description_
    &#34;&#34;&#34;
    features = self.extract_all_features(&#34;CDS&#34;, None)
    cds_seq_records: List[SeqRecord] = []
    for idx, feature in enumerate(features, 1):
        qualifiers = feature.qualifiers
        protein_id = qualifiers.get(&#34;protein_id&#34;, [None])[0]
        product = qualifiers.get(&#34;product&#34;, [&#34;&#34;])[0]
        translation = qualifiers.get(&#34;translation&#34;, [None])[0]

        start = self._to_int(feature.location.start)
        end = self._to_int(feature.location.end)
        strand = &#34;+&#34; if feature.strand == 1 else &#34;-&#34;

        location_id = f&#34;|{start}_{end}_{strand}|&#34;
        if protein_id is None:
            seq_id = f&#34;GENE{idx:06d}{location_id}&#34;
        else:
            seq_id = f&#34;GENE{idx:06d}_{protein_id}{location_id}&#34;

        cds_seq_record = SeqRecord(
            seq=Seq(translation), id=seq_id, description=product
        )
        cds_seq_records.append(cds_seq_record)

    SeqIO.write(cds_seq_records, fasta_outfile, &#34;fasta-2line&#34;)</code></pre>
</details>
</dd>
<dt id="mgcplotter.genbank.Genbank.write_genome_fasta"><code class="name flex">
<span>def <span class="ident">write_genome_fasta</span></span>(<span>self, outfile: Union[str, pathlib.Path]) ‑> None</span>
</code></dt>
<dd>
<div class="desc"><p>Write genome fasta file</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>outfile</code></strong> :&ensp;<code>Union[str, Path]</code></dt>
<dd>Output genome fasta file</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def write_genome_fasta(
    self,
    outfile: Union[str, Path],
) -&gt; None:
    &#34;&#34;&#34;Write genome fasta file

    Args:
        outfile (Union[str, Path]): Output genome fasta file
    &#34;&#34;&#34;
    write_seq = self.genome_seq
    with open(outfile, &#34;w&#34;) as f:
        f.write(f&#34;&gt;{self.name}\n{write_seq}\n&#34;)</code></pre>
</details>
</dd>
</dl>
</dd>
</dl>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="mgcplotter" href="index.html">mgcplotter</a></code></li>
</ul>
</li>
<li><h3><a href="#header-classes">Classes</a></h3>
<ul>
<li>
<h4><code><a title="mgcplotter.genbank.Genbank" href="#mgcplotter.genbank.Genbank">Genbank</a></code></h4>
<ul class="">
<li><code><a title="mgcplotter.genbank.Genbank.gc_skew" href="#mgcplotter.genbank.Genbank.gc_skew">gc_skew</a></code></li>
<li><code><a title="mgcplotter.genbank.Genbank.gc_content" href="#mgcplotter.genbank.Genbank.gc_content">gc_content</a></code></li>
<li><code><a title="mgcplotter.genbank.Genbank.extract_all_features" href="#mgcplotter.genbank.Genbank.extract_all_features">extract_all_features</a></code></li>
<li><code><a title="mgcplotter.genbank.Genbank.write_cds_fasta" href="#mgcplotter.genbank.Genbank.write_cds_fasta">write_cds_fasta</a></code></li>
<li><code><a title="mgcplotter.genbank.Genbank.write_genome_fasta" href="#mgcplotter.genbank.Genbank.write_genome_fasta">write_genome_fasta</a></code></li>
</ul>
</li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc" title="pdoc: Python API documentation generator"><cite>pdoc</cite> 0.10.0</a>.</p>
</footer>
</body>
</html>